### PR TITLE
ENH: Implement PeriodIndex.intersection without object-dtype cast

### PR DESCRIPTION
**Pull request by jbrockmendel**
PeriodIndex._simple_new is not as simple as it should be.  In order to get it to have the appropriate signature we need to implement some of the set methods correctly.  This is the first of those.

### CODE REVIEW

**Review by TomAugspurger — APPROVED** [review]
LGTM. How does the ASV for `period.Indexing.time_intersection` look?

**Comment by jbrockmendel** [comment]
> How does the ASV for period.Indexing.time_intersection look?

For the length=1000 index in the benchmark we take a hit of about 37%.  When I change the benchmark to be length=10**5 it is no change.

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -799,6 +803,30 @@ def _wrap_setop_result(self, other, result):
         result.name = name
         return result
 
+    def intersection(self, other, sort=False):
```
Comment (jreback) [inline pandas/core/indexes/period.py]: type at some point

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -799,6 +803,30 @@ def _wrap_setop_result(self, other, result):
...
+        self._validate_sort_keyword(sort)
+        self._assert_can_do_setop(other)
+        res_name = get_op_result_name(self, other)
+        other = ensure_index(other)
+
```
Comment (jreback) [inline pandas/core/indexes/period.py]: can this code be shared with superclasses?

**Review by jreback — APPROVED** [review]
lgtm. some comments for followon.

**Comment by jreback** [comment]
> > How does the ASV for period.Indexing.time_intersection look?
> 
> For the length=1000 index in the benchmark we take a hit of about 37%. When I change the benchmark to be length=10**5 it is no change.

interesting, i though this would be faster. ok

File: pandas/core/indexes/period.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: i think so, its going to be a few steps before we get there

**Pull request merged**